tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionParser.cs

```diff
@@ -0,0 +1,483 @@
+{
+    private static readonly LabelTarget ReturnTarget = Expression.Label(typeof(T));
+    private static readonly Func<ScopeMember, ScopeMember[], T> DefaultDelegate;
+
+    private List<EvaluationError> _errors;
+    private int _arrayStack;
+
+    static ProbeExpressionParser()
+    {
+        DefaultDelegate = (_, _) =>
+        {
+            if (typeof(T) == typeof(bool))
+            {
+                return (T)(object)true;
+            }
+
+            return default;
+        };
+    }
+
+    private delegate Expression Combiner(Expression left, Expression right);
+
+    private Expression ParseRoot(
+        JsonTextReader reader,
+        List<ParameterExpression> parameters)
+    {
+        var readerValue = reader.Value?.ToString();
+        switch (reader.TokenType)
+        {
+            case JsonToken.PropertyName:
+                switch (readerValue)
+                {
+                    case "and":
+                        {
+                            return ConditionalOperator(reader, Expression.AndAlso, parameters);
+                        }
+
+                    case "or":
+                        {
+                            return ConditionalOperator(reader, Expression.OrElse, parameters);
+                        }
+
+                    default:
+                        return ParseTree(reader, parameters, null, false);
+                }
+        }
+
+        return null;
+    }
+
+    private Expression ConditionalOperator(JsonTextReader reader, Combiner combiner, List<ParameterExpression> parameters)
+    {
+        Expression Combine(Expression leftOperand, Expression rightOperand) =>
+            leftOperand == null ? rightOperand : combiner(leftOperand, rightOperand.Type != ProbeExpressionParserHelper.UndefinedValueType ? rightOperand : Expression.Constant(true));
+
+        _arrayStack++;
+        reader.Read();
+        var right = ParseTree(reader, parameters, null);
+        var left = Combine(null, right);
+
+        while (reader.Read())
+        {
+            if (reader.TokenType == JsonToken.StartArray)
+            {
+                _arrayStack++;
+                continue;
+            }
+            else if (reader.TokenType == JsonToken.EndArray)
+            {
+                _arrayStack--;
+                continue;
+            }
+
+            if (reader.TokenType is JsonToken.StartObject or JsonToken.EndObject)
+            {
+                continue;
+            }
+
+            if (_arrayStack == 0)
+            {
+                break;
+            }
+
+            right = ParseTree(reader, parameters, null, false);
+            left = Combine(left, right);
+        }
+
+        return left;
+    }
+
+    private Expression ParseTree(
+        JsonTextReader reader,
+        List<ParameterExpression> parameters,
+        ParameterExpression itParameter,
+        bool shouldAdvanceReader = true)
+    {
+        while (ConditionalRead(reader, shouldAdvanceReader))
```

Comment:
It always concerns me seeing while statements with opaque expression that is not fully controlled.
What do you think about adding a safeguard around it to make sure we will _never_ fall into an infinite loop?